// eoslime/.history/src/network-providers/provider_20191028104858.js
const BosProvider = require('./bos-provider')
const MainProvider = require('./main-provider')
const KylinProvider = require('./kylin-provider')
const LocalProvider = require('./local-provider')
const JungleProvider = require('./jungle-provider')
const WorbliProvider = require('./worbli-provider')
const CustomProvider = require('./custom-provider')

const NETWORKS = {
    bos: () => { return new BosProvider() },
    main: () => { return new MainProvider() },
    kylin: () => { return new KylinProvider() },
    local: () => { return new LocalProvider() },
    jungle: () => { return new JungleProvider() },
    worbli: () => { return new WorbliProvider() },
    custom: (networkConfig) => { return new CustomProvider(networkConfig) }
}

static createFunction(contract, functionName, contractStructs) {
    const contractFunction = new ContractFunction(contract, functionName, contractStructs);

    const proxyHandler = {
        get: (obj, value) => {
            if (!obj.hasOwnProperty(value)) {
                return contractFunction[value];
            }
        }
    }

    const proxy = new Proxy(async function (...params) {
        return contractFunction.broadcast(...params);
    }, proxyHandler);

    return proxy;
}


class Provider {
    constructor(network) {
        if (NETWORKS[network]) {
            return NETWORKS[network]();
            // provider.init = this.build;
            // provider.NETWORKS = Provider.availableNetworks().all;
        }

        return NETWORKS.custom(network);
        // provider.init = this.build;
        // provider.NETWORKS = Provider.availableNetworks().all;
    }

    build(network) {
        return new Provider(network);
    }

    static availableNetworks() {
        const networks = {};
        Object.keys(NETWORKS).forEach(networkName => {
            networks[networkName.toUpperCase] = networkName;
        });

        networks.all = Object.keys(NETWORKS);
        return networks;
    }
}

module.exports = Provider
