// eoslime/.history/src/network-providers/providers-factory_20191028133252.js
const is = require('./../helpers/is');
const Provider = require('./provider');
const BaseProvider = require('./base-provider');

class ProvidersFactory {

    create(network) {
        this.provider = constructProvider(network);

        const providerProxyHandler = {
            get: (obj, value) => {
                if (!obj.hasOwnProperty(value)) {
                    return this.provider[value];
                }

                return obj[value];
            },
            construct: (target, network) => {
                return constructProvider(network);
            }
        }

        return new Proxy(P, providerProxyHandler);
    }

    reset(newProvider) {
        is(newProvider).instanceOf(BaseProvider);
        Object.assign(this.provider, newProvider);
    }

    static availableNetworks() {
        const networks = {};
        Object.keys(NETWORKS).forEach(networkName => {
            networks[networkName.toUpperCase] = networkName;
        });

        networks.all = Object.keys(NETWORKS);
        return networks;
    }
}

const constructProvider = function (network) {
    if (NETWORKS[network]) {
        return NETWORKS[network]();
    }

    return NETWORKS.custom(network);
}

module.exports = ProvidersFactory;
