// eoslime/.history/src/network-providers/provider_20191028104959.js
const BosProvider = require('./bos-provider')
const MainProvider = require('./main-provider')
const KylinProvider = require('./kylin-provider')
const LocalProvider = require('./local-provider')
const JungleProvider = require('./jungle-provider')
const WorbliProvider = require('./worbli-provider')
const CustomProvider = require('./custom-provider')

const NETWORKS = {
    bos: () => { return new BosProvider() },
    main: () => { return new MainProvider() },
    kylin: () => { return new KylinProvider() },
    local: () => { return new LocalProvider() },
    jungle: () => { return new JungleProvider() },
    worbli: () => { return new WorbliProvider() },
    custom: (networkConfig) => { return new CustomProvider(networkConfig) }
}

class Provider {
    constructor(network) {
        const providerProxyHandler = {
            get: (obj, value) => {
                if (!obj.hasOwnProperty(value)) {
                    return provider[value];
                }
            }
        }

        const providerProxy = new Proxy(async function (...params) {
            return contractFunction.broadcast(...params);
        }, proxyHandler);

        return proxy;


        if (NETWORKS[network]) {
            const provider = NETWORKS[network]();

            // provider.init = this.build;
            // provider.NETWORKS = Provider.availableNetworks().all;
        }

        return NETWORKS.custom(network);
        // provider.init = this.build;
        // provider.NETWORKS = Provider.availableNetworks().all;
    }

    build(network) {
        return new Provider(network);
    }

    static availableNetworks() {
        const networks = {};
        Object.keys(NETWORKS).forEach(networkName => {
            networks[networkName.toUpperCase] = networkName;
        });

        networks.all = Object.keys(NETWORKS);
        return networks;
    }
}

module.exports = Provider
