// eoslime/.history/src/account/base-account_20191015094307.js
const is = require('../helpers/is')
const eosECC = require('eosjs').modules.ecc;

class BaseAccount {
    constructor(name, privateKey, provider, permission) {
        this.name = name;
        this.provider = provider;
        this.executiveAuthority = {
            actor: name,
            permission: permission
        }

        this.privateKey = privateKey;
        this.publicKey = eosECC.PrivateKey.fromString(privateKey).toPublic().toString();
    }

    addProposerAccount(account) {
        // this.proposers.push({
        //     actor: account.executiveAuthority.actor,
        //     permission: account.executiveAuthority.permission
        // })
    }

    async propose(action, proposer) {
        // const accountInfo = await this.getAuthorityInfo();
        // const proposalData =
        // {
        //     proposer: proposer,
        //     proposal_name: action,
        //     requested: [
        //         {
        //             actor: proposer.actor,
        //             permission: proposer.permission
        //         },
        //         {
        //             actor: 'bob',
        //             permission: 'active'
        //         }
        //     ]
        // }

        // await this.provider.eos.transact({
        //     actions: [{
        //         account: 'eosio.msig',
        //         name: 'propose',
        //         authorization: [{
        //             actor: proposer.actor,
        //             permission: proposer.permission,
        //         }],
        //         data: proposalData,
        //     }]
        // });
    }

    approveProposal() {

    }

}

module.exports = BaseAccount
