// eoslime/.history/src/account/multi-signature-account_20191025112552.js
const is = require('../helpers/is')
const Account = require('./account');
const ContractFactory = require('./../contract/contract-factory');

class MultiSignatureAccount extends Account {

    constructor(name, privateKey, provider, authority) {
        super(name, privateKey, provider, authority)

        this.accounts = [];
        this.proposals = {};
        this.proposalsActions = {};
    }

    async buyRam(bytes, payer = this) {
        is(payer).instanceOf(Account);

        const contractFactory = new ContractFactory(this.provider);
        const systemContract = await contractFactory.at('system');
        return this.propose(systemContract.buyram, [
            payer.name,
            this.name,
            bytes
        ]);
    }

    async buyBandwidth(cpu, net, payer = this) {
        is(payer).instanceOf(Account);

        return this.provider.eos.transaction(tr => {
            tr.delegatebw({
                from: payer.name,
                receiver: this.name,
                stake_cpu_quantity: cpu,
                stake_net_quantity: net,
                transfer: 0
            });
        }, { keyProvider: payer.privateKey });
        this.propose();
    }

    async send(receiver, amount, symbol = 'EOS') {
        is(receiver).instanceOf(Account);

        return this.provider.eos.transfer(
            this.name,
            receiver.name,
            `${amount} ${symbol}`,
            this.executiveAuthority,
            { broadcast: true, sign: true, keyProvider: this.privateKey }
        );
        this.propose();
    }

    async createAuthority(authorityName, threshold = 1) {
        const authorization = {
            threshold,
            keys: [{ key: this.publicKey, weight: threshold }]
        }
        await setAuthority.call(this, authorityName, this.executiveAuthority.permission, authorization);

        return new Account(this.name, this.privateKey, this.provider, authorityName);
        this.propose();
    }

    async setAuthorityAbilities(abilities) {
        is(abilities).instanceOf(Array);

        /*
            abilities => 
                [
                    action,
                    contract
                ]
            
        */

        await this.provider.eos.transaction(tr => {
            for (let i = 0; i < abilities.length; i++) {
                const ability = abilities[i];
                tr.linkauth({
                    account: this.name,
                    code: ability.contract,
                    type: ability.action,
                    requirement: authorityName
                }, { authorization: [this.executiveAuthority] });
            }
        }, { broadcast: true, sign: true, keyProvider: this.privateKey });
        this.propose();
    }

    async setThreshold(threshold) {
        this.propose();
    }

    async addPermission(authorityName, weight = 1) {
        return this.addOnBehalfAccount(this.name, authorityName, weight);
    }

    async addOnBehalfAccount(accountName, authorityName, weight = 1) {
        this.propose();
    }

    async addAuthorityKey(publicKey, weight = 1) {
        this.propose();
    }

    async setWeight() {
        this.propose();
    }

    loadKeys(privateKeys) {
        for (let i = 0; i < privateKeys.length; i++) {
            this.accounts.push(new Account(this.name, privateKeys[i], this.provider, this.executiveAuthority.permission));
        }
    }

    loadAccounts(accounts) {
        for (let i = 0; i < accounts.length; i++) {
            is(accounts[i]).instanceOf(Account);
            this.accounts.push(accounts[i]);
        }
    }

    async propose(contractAction, actionData) {
        const actionTx = await contractAction.sign(this, ...actionData);
        const proposalId = Date.now();

        this.proposals[proposalId] = actionTx;
        this.proposalsActions[proposalId] = { action: contractAction, data: actionData };

        return proposalId;
    }

    async approve(approver, proposalId) {
        is(approver).instanceOf(Account);
        requireExistingProposal(this.proposals, proposalId);

        const proposalAction = this.proposalsActions[proposalId];

        let actionTx;
        if (hasTheProposerName(approver, this)) {
            actionTx = await proposalAction.action.sign(approver, ...proposalAction.data);
        } else {
            actionTx = await proposalAction.action.signOnBehalf(approver, this, ...proposalAction.data);
        }

        this.proposals[proposalId].signatures.push(actionTx.signatures[0]);
    }

    async approveAll(proposalId) {
        for (let i = 0; i < this.accounts.length; i++) {
            await this.approve(this.accounts[i], proposalId);
        }
    }

    async processProposal(proposalId) {
        requireExistingProposal(this.proposals, proposalId);
        requireEnoughApprovals(this, this.proposals[proposalId]);

        const proposalTx = this.proposals[proposalId];
        delete this.proposals[proposalId];
        delete this.proposalsActions[proposalId]

        return this.provider.eos.pushTransaction(proposalTx);
    }
}

const requireExistingProposal = function (proposals, proposalId) {
    if (!proposals[proposalId]) {
        throw new Error('Such proposal does not exists');
    }
}

const requireEnoughApprovals = async function (account, proposal) {
    const authorityInfo = await account.getAuthorityInfo();
    if (authorityInfo.required_auth.threshold != proposal.signatures.length) {
        throw new Error(`Proposal should be approved minimum ${authorityInfo.required_auth.threshold - 1} times, but it is approved ${proposal.signatures.length - 1}`);
    }
}

const hasTheProposerName = function (candidate, proposer) {
    return candidate.name == proposer.name
}

module.exports = MultiSignatureAccount
