// eoslime/.history/src/account/authrotiy-account/account_20191114135003.js
const is = require('../../helpers/is')
const NormalAccount = require('./../normal-account/account');
console.log(NormalAccount)

class AuthorityAccount extends NormalAccount {

    constructor(name, privateKey, provider, permission, parentPermission) {
        super(name, privateKey, provider, permission);
        this.parentPermission = parentPermission;
    }

    async setAuthorityAbilities(abilities) {
        is(abilities).instanceOf('Array');

        await this.provider.eos.transaction(tr => {
            for (let i = 0; i < abilities.length; i++) {
                const ability = abilities[i];
                tr.linkauth({
                    account: this.name,
                    code: ability.contract,
                    type: ability.action,
                    requirement: this.executiveAuthority.permission
                }, { authorization: [`${this.name}@${this.parentPermission}`] });
            }
        }, { broadcast: true, sign: true, keyProvider: this.privateKey });
    }
}

module.exports = AuthorityAccount;
