// eoslime/.history/src/account/authrotiy-account/authority-account-factory_20191114140030.js
const AuthorityAccount = require('./account');

class AuthorityAccountFactory {

    static create(name, privateKey, provider, permission, parentPermission) {
        return new AuthorityAccount(name, privateKey, provider, permission, parentPermission);
    }

    async setAuthorityAbilities(account, parentAuthority) {
        return async function (abilities) {
            is(abilities).instanceOf('Array');

            await account.provider.eos.transaction(tr => {
                for (let i = 0; i < abilities.length; i++) {
                    const ability = abilities[i];
                    tr.linkauth({
                        account: account.name,
                        code: ability.contract,
                        type: ability.action,
                        requirement: account.executiveAuthority.permission
                    }, { authorization: [`${this.name}@${parentAuthority}`] });
                }
            }, { broadcast: true, sign: true, keyProvider: account.privateKey });
        }
    }
}

module.exports = AuthorityAccountFactory;
