// eoslime/.history/src/account/multi-signature-account/account_20191111102731.js
const is = require('../../helpers/is')
const Account = require('../normal-account/account');

class MultiSignatureAccount extends Account {

    constructor(name, privateKey, provider, authority) {
        super(name, privateKey, provider, authority)

        this.accounts = [];
        this.proposals = {};
        this.proposalsActions = {};
    }

    loadKeys(privateKeys) {
        for (let i = 0; i < privateKeys.length; i++) {
            this.accounts.push(new Account(this.name, privateKeys[i], this.provider, this.executiveAuthority.permission));
        }
    }

    loadAccounts(accounts) {
        for (let i = 0; i < accounts.length; i++) {
            is(accounts[i]).instanceOf(Account);
            this.accounts.push(accounts[i]);
        }
    }

    async propose(contractAction, actionData) {
        const actionTx = await contractAction.sign(this, ...actionData);
        const proposalId = Date.now();

        this.proposals[proposalId] = actionTx;
        this.proposalsActions[proposalId] = { action: contractAction, data: actionData };

        return proposalId;
    }

    async approve(approver, proposalId) {
        is(approver).instanceOf(Account);
        requireExistingProposal(this.proposals, proposalId);

        const proposalAction = this.proposalsActions[proposalId];

        console.log(approver.name)
        console.log(tnis.name)
        if (hasTheProposerName(approver, this)) {
            throw new Error('Proposer has already signed the transaction');
        } else {
            const actionTx = await proposalAction.action.signOnBehalf(approver, this, ...proposalAction.data);
            this.proposals[proposalId].signatures.push(actionTx.signatures[0]);
        }
    }

    async approveAll(proposalId) {
        for (let i = 0; i < this.accounts.length; i++) {
            await this.approve(this.accounts[i], proposalId);
        }
    }

    async processProposal(proposalId) {
        requireExistingProposal(this.proposals, proposalId);
        requireEnoughApprovals(this, this.proposals[proposalId]);

        const proposalTx = this.proposals[proposalId];
        delete this.proposals[proposalId];
        delete this.proposalsActions[proposalId]

        return this.provider.eos.pushTransaction(proposalTx);
    }
}

const requireExistingProposal = function (proposals, proposalId) {
    if (!proposals[proposalId]) {
        throw new Error('Such proposal does not exists');
    }
}

const requireEnoughApprovals = async function (account, proposal) {
    const authorityInfo = await account.getAuthorityInfo();
    if (authorityInfo.required_auth.threshold != proposal.signatures.length) {
        throw new Error(`Proposal should be approved minimum ${authorityInfo.required_auth.threshold - 1} times, but it is approved ${proposal.signatures.length - 1}`);
    }
}

const hasTheProposerName = function (candidate, proposer) {
    return candidate.name == proposer.name
}

module.exports = MultiSignatureAccount
