// eoslime/.history/cli-commands/test/test-frameworks/mocha/index_20190918185746.js
const Mocha = require('mocha');
const mochaConfig = require('./mocha-config');

const EVENTS = {
    'allFinished': 'suite end'
}

class MochaFramework {
    constructor() {
        this.instance = new Mocha(mochaConfig);
        this.eventsHooks = [];

    }

    addTestFiles(testFiles) {
        testFiles.forEach(testFile => {
            this.instance.addFile(testFile);
        });
    }

    on(eventName, callback) {
        this.eventsHooks.push({
            event: EVENTS[eventName],
            callback: callback
        });
    }

    async runTests(...params) {
        const self = this;
        return new Promise((resolve, reject) => {

            const mochaRunner = self.instance.run(
                failures => {
                    process.exitCode = failures ? -1 : 0;
                    if (failures) {
                        return void reject(failures)
                    }

                    resolve();
                }
            );

            self.instance.suite.on('pre-require', function (context) {
                const origDescribe = context.describe;
                console.log(context)
                context.describe = function (name, impl) {
                    return origDescribe.call(this, name, function () {
                        console.log("here")
                    });
                };
            });

            for (let i = 0; i < self.eventsHooks.length; i++) {
                const eventHook = self.eventsHooks[i];
                mochaRunner.on(eventHook.event, eventHook.callback);
            }
        });
    }
}

module.exports = MochaFramework;
