// eoslime/.history/cli-commands/test/options/resource-usage-option/report-table_20191003154804.js
const chalk = require('chalk').default;
const CLITable = require('cli-table');

const TABLE_HEAD = {
    head: ['', 'Contract', 'Action', 'CPU ( MIN | MAX )', 'NET ( MIN | MAX )', 'RAM ( MIN | MAX )', 'Calls']
}

class ReportTable {
    constructor() {
        this.table = new CLITable(TABLE_HEAD);
    }

    addRow(rowData) {
        rowData[3] = `${rowData[3][0]} μs | ${rowData[3][1]} μs`;
        rowData[4] = `${rowData[4][0]} Bytes | ${rowData[4][1]} Bytes`;
        rowData[5] = `${rowData[5][0]} Bytes ${rowData[5][1]} Bytes`;
        this.table.push(rowData);
    }

    addSection(sectionName, sectionRows) {
        this.table.push({ [chalk.cyanBright(sectionName)]: ['', '', '', '', '', ''] });

        if (sectionRows) {
            for (let i = 0; i < sectionRows.length; i++) {
                const sectionRow = sectionRows[i];
                sectionRow.unshift('');
                this.addRow(sectionRow);
            }
        }
    }

    draw() {
        console.log(this.table.toString());
    }
}

module.exports = ReportTable;
