// eoslime/.history/index_20191028122417.js
const Provider = require('./src/network-providers/provider');

const AccountFactory = require('./src/account/account-factory');
const ContractFactory = require('./src/contract/contract-factory');

const utils = require('./src/utils');

module.exports = (function () {

    const init = function (network = 'local') {
        const provider = new Provider(network);
        // const providerProxyHandler = {
        //     get: (obj, value) => {
        //         if (!obj.hasOwnProperty(value)) {
        //             return provider[value];
        //         }
        //     },
        //     construct: (target, network) => {
        //         return new target(network);
        //     }
        // }

        const accountFactory = new AccountFactory(provider);
        const contractFactory = new ContractFactory(provider);

        return {
            utils: utils,
            Provider: new Proxy(Provider, providerProxyHandler),
            Account: accountFactory,
            Contract: contractFactory
        };
    }

    return {
        init: init,
        utils: utils,
        NETWORKS: Provider.availableNetworks().all
    };
})();
